pkg/destroy/aws/aws.go

```diff
@@ -1825,6 +1825,41 @@ func deleteS3(session *session.Session, arn arn.ARN, logger logrus.FieldLogger)
 	if err != nil && !isBucketNotFound(err) {
 		return err
 	}
+
+	var lastError error
+	err = client.ListObjectVersionsPagesWithContext(aws.BackgroundContext(), &s3.ListObjectVersionsInput{
+		Bucket: aws.String(arn.Resource)}, func(page *s3.ListObjectVersionsOutput, lastPage bool) bool {
+		var deleteObjects []*s3.ObjectIdentifier
+		for _, deleteMarker := range page.DeleteMarkers {
+			deleteObjects = append(deleteObjects, &s3.ObjectIdentifier{
+				Key:       aws.String(*deleteMarker.Key),
+				VersionId: aws.String(*deleteMarker.VersionId),
+			})
+		}
+		for _, version := range page.Versions {
+			deleteObjects = append(deleteObjects, &s3.ObjectIdentifier{
+				Key:       aws.String(*version.Key),
+				VersionId: aws.String(*version.VersionId),
+			})
+		}
+		_, err := client.DeleteObjects(&s3.DeleteObjectsInput{
```

Comment:
Is there a limit as to how many objects can be deleted in one API call?

Because we could collect all the deleteObjects in this list pages block and later on delete them all in one API call instead of calling delete per page..

Reply:
Also if we look at line 1821, right above..

```
err := s3manager.NewBatchDeleteWithClient(client).Delete(aws.BackgroundContext(), iter)
	if err != nil && !isBucketNotFound(err) {
		return err
	}
```

It takes an iterator.. maybe we can collect all the deletemarker and version objects, create an iterator for those and then call batch delete?

Reply:
Yes, there is a 1000 object limit when using deleteObjects() (which is what is being used here). This code already does what you are requesting. ListObjectVersionsPagesWithContext() will return a page with 1000 objects, these are gathered into deleteObjects, which are then deleted with DeleteObjects(). I did not see anything in s3manager that could be used here. I'll take a deeper look and see if I can find anything.

Reply:
@jhixson74 
> Yes, there is a 1000 object limit when using deleteObjects() (which is what is being used here). 
can you point to the aws docs that provide that information?

> This code already does what you are requesting. ListObjectVersionsPagesWithContext() will return a page with 1000 objects,

I don't see you setting the page item limit here, can you provide link to doc that provides information about this default page item size.

also it looks like for there are 2 items _delete marker_ and _version object_ per item in one page, wouldn't that affect how many we are sending in one deleteobjects api call.

Reply:
I'll dig up the documentation. Yes, delete marker + version make up the 1000. If there are 2 delete markers, then there are 998 version objects. If there are 20 delete markers, there are 980 delete markers. 1000 is always what is returned, and 1000 is the delete maximum.

Reply:
DeleteObjects: https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.DeleteObjects

Reply:
I am still trying to find the docs for ListObjectVersionsPagesWithContext. I definitely confirmed it through a small driver program before adding it though. There is a page max of 1000.... somewhere... deep... inside of aws docs.

Reply:
I can't seem to find it. I had closed out all my documentation windows when I pushed this. Nevertheless, I've set it explicitly here now.

Reply:
https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectVersions.html#API_ListObjectVersions_RequestSyntax

```
 max-keys

    Sets the maximum number of keys returned in the response. By default the API returns up to 1,000 key names. The response might contain fewer keys but will never contain more. If additional keys satisfy the search criteria, but were not returned because max-keys was exceeded, the response contains <isTruncated>true</isTruncated>. To return the additional keys, see key-marker and version-id-marker.
```